### ISSUE TITLE
Make the site load faster

### ISSUE THREAD

**Issue by cattsmall** [QuantumBox/blackgamedevs#122]
How might we make the page load faster while also making it easier to edit entries and scale the list?

- [x] SSH certificate
- [x] Modify JS to only update to the DOM once (rather than for each entry)
- [x] Move to Gatsby

**Comment by MaximumCrash** [comment]
Howdy @cattsmall, I'll take a look through the codebase and see how I can help.

**Comment by MaximumCrash** [comment]
@cattsmall So my current audit is that we have a good problem. From my observation, the site is loading slowly because there's over 200+ entries of game devs (not including companies) and the scripts.js is recursively generating filter properties for every single dev/company including internal lists (like Skills). The more devs we add, the slower the site becomes with this current solution.

The best course of action from here would be to begin moving the platform to a static site generator like Gatsby, and then serving it on Netlify. This will turbo charge the site to have the data baked during build time, instead of every time someone goes to the website.

 Since I know we're all strapped for time, I can begin the site rebuild in Gatsby, and perform bits of the migration as I go. I hope to make the site blazingly fast, but also easier for the team to maintain and for folks to add/edit entries.

**Comment by MaximumCrash** [comment]
@cattsmall So my current audit is that we have a good problem. From my observation, the site is loading slowly because there's over 200+ entries of game devs (not including companies) and the scripts.js is recursively generating filter properties and updating the DOM for every single dev/company including internal lists (like Skills). The more devs we add, the slower the site becomes with this current solution.

The quickest solution would be to just update the DOM once, however this might only be a band-aid since people who go to the site would be loading the content client side every time they visit (there isn't any caching from what I could tell). 

I believe the best course of action from here could be to begin moving the platform to a static site generator like Gatsby, and then serving it on Netlify. This will turbo charge the site to have the data baked during build time, instead of every visit.

Since I know we're all strapped for time, I can begin the site rebuild in Gatsby, and perform bits of the migration as I go. I hope to make the site blazingly fast, but also easier for the team to maintain and for folks to add/edit entries. Let me know if this is cool

**Comment by cattsmall** [comment]
I seeeee! Let me look into the benefits and costs and I'll get back to ya later today!

**Comment by cattsmall** [comment]
OK! In the short term I've updated the code to not be so hacky and instead update the DOM once. 

@MaximumCrash I looked into things and it seems like Netlify costs money. Why not use Gatsby with GitHub Pages, for example? Tell me your thoughts!

**Comment by MaximumCrash** [comment]
Awesome it's wayyy faster!

I honestly don't have an argument against using Gatsby with GitHub Pages. I just happen have more history with Netlify 😅.  I'm totally down to go with Gatsby + GH Pages @cattsmall.

**Comment by cattsmall** [comment]
Sounds good, let's do it! LMK how I can help; I closed out any addition requests so all the content should be up-to-date.

**Comment by cattsmall** [comment]
Hey @MaximumCrash! @annarankin and @CentroDL have expressed helping with this as well.

Two questions:
- Where are you right now with that there Gatsby?
- Is there a way to divide and conquer?

**Comment by MaximumCrash** [comment]
> Hey @MaximumCrash! @annarankin and @CentroDL have expressed helping with this as well.
> 
> Two questions:
> 
> * Where are you right now with that there Gatsby?
> * Is there a way to divide and conquer?

@annarankin @CentroDL I apologize for missing this. I've been having a case of tunnel vision and have been missing my github notifications lately. I'd still love to sync if you have time. My first pass has been completed on a branch called rejon/BGD_v2. I'm still in the process of documenting, but you can view a live example thus far here: https://maximumcrash.github.io/blackgamedevs_v2/

**Comment by zelgadis** [comment]
@MaximumCrash Looks great! Just wanted to chime in and say that I'm interested in helping out as well! I'd love to connect with either you or @cattsmall on how I can contribute!

**Comment by MaximumCrash** [comment]
@zelgadis Thank you. I believe it would be up to @cattsmall and the rest of the team to make the call on where we take this next.

**Comment by annarankin** [comment]
Hiya @MaximumCrash -- this is awesome! I'd love to help get this in and I know @cattsmall 's excited about it as well!). What's the best way for me to help? I can offer code review if that's needed. I've been a bit busy with day-job-stuff, but feel free to loop me in wherever you like 😄

**Comment by MaximumCrash** [comment]
> Hiya @MaximumCrash -- this is awesome! I'd love to help get this in and I know @cattsmall 's excited about it as well!). What's the best way for me to help? I can offer code review if that's needed. I've been a bit busy with day-job-stuff, but feel free to loop me in wherever you like 😄

Thank you 😊  

I'd love a code review, the more folks understand the underlying bits the better. There are still a number of things I need to document. I'm hoping to provide a handbook on my thinking, the stack, why things were done a certain way ect, by the end of this month. If you have any heavy questions while you're looking through those please let me know 🙏

**Comment by zelgadis** [comment]
@MaximumCrash I'll give it a look-over as well!

**Comment by MaximumCrash** [comment]
@zelgadis Please go right ahead. I won't stop ya 😁

**Comment by zelgadis** [comment]
(For posterity, here is the link to a diff: https://github.com/QuantumBox/blackgamedevs/compare/rejon/siteV2)

**Comment by zelgadis** [comment]
@MaximumCrash No pressure / hurry, but if you make a pull request for your branch, that'll make it easier to code review! Right now I can only make comments on the individual commits, rather than the piece as a whole.

(I tried to create a PR off your branch, but github wouldn't let me)

**Comment by zelgadis** [comment]
@MaximumCrash I added some feedback! I kept my feedback focused more on the output than the code, for a few reasons:
1. I'm not much of a front-end developer 🤓
2. With open-source projects I lean more towards results and unblocking collaboration than on scrutinizing the code
3. Without a formalized pull request its harder to comment on code changes, since you can only add comments to an individual commit (afaict)

Anyways, thanks so much for putting together v2 of this site! I hope my comments are helpful! 😅

**Comment by MaximumCrash** [comment]
@zelgadis thank you for going through! I'll respond to your notes in tandem here:

1. Ah yes building the project is part of the documentation I'm doing for the complete hand off. 

2. I agree. I've currently minimized all the images that existed as much as possible to improve site speed. Pulling back on the minimizer for the sake quality is now possible more so than ever because of our strategy with pagination. I'm only a little reserved on pumping up the quality to the max because this implementation is meant to be a direct migration without a loss of content (my run-once transformer script downloads the images directly and throws them in the directory folder instead of utilizing the old strategy of folks providing a cross-site link to their image).

3. I also agree here. I've already updated the transformer script on my end to move them into the static directory. Moreso because Gatsby w/ MDX has a tendency to be inconsistent with it's image loading if they aren't directly available in the public directory. (all content in static is put in the public directory on build, the public directory is what gets deployed for gh-pages)

4. I disagree here. 

Here's the TLDR: I did this on purpose to decouple differences between files; with the goal of creating less friction when having to add a new entry (it also leans into my philosophy on WYSIWYG with MDX). Another added benefit is that it's less prone to human errors. Best to have the data dictate the rendering instead of the directory location.

My line of thinking was to not have the individual files seperated into contextual usage. While it seems more organized to seperate the directories this gets complicated once you want to add a new type of entry. My aim was to decouple the need for different kinds of directories, and different kinds of files, by leveraging the frontmatter yaml boolean "isCompany" to ease builds times and make it easier to build solutions. For example, instead of running a GraphQL query for each individual directory we'd need to support we can instead run a filter to  on the frontmatter. Another example is how the search indexes are built. Maintaining individual directories would create friction for devlelopers (and normal folks) to have to remember to include specific directories. 

In conclusion, we can avoid conflicts (especially merge conflicts) if we don't allow the opportunity for them to happen. To each their own file. 

5. This is a good idea. I will add it in. 

I've also opened a PR here (but I doubt it's pointing at the right branch): https://github.com/QuantumBox/blackgamedevs/pull/215
